Mathlib/NumberTheory/LSeries/AbstractFuncEq.lean

```diff
@@ -0,0 +1,471 @@
+
+lemma symm_Λ_eq : P.symm.Λ = mellin P.g := by rfl
+
+/-- If `(f, g)` are a strong FE pair, then the Mellin transform of `f` is entire. -/
+lemma differentiable_Λ : Differentiable ℂ P.Λ := fun s ↦
+  let ⟨_, ht⟩ := exists_gt s.re
+  let ⟨_, hu⟩ := exists_lt s.re
+  mellin_differentiableAt_of_isBigO_rpow P.hf_int (P.hf_top' _) ht (P.hf_zero' _) hu
+
+/-- Main theorem about strong FE pairs: if `(f, g)` are a strong FE pair, then the Mellin
+transforms of `f` and `g` are related by `s ↦ k - s`.
+
+This is proved by making a substitution `t ↦ t⁻¹` in the Mellin transform integral. -/
+lemma functional_equation (s : ℂ) :
+    P.Λ (P.k - s) = P.ε • P.symm.Λ s := by
+  -- unfold definition:
+  rw [P.Λ_eq, P.symm_Λ_eq]
+  -- substitute `t ↦ t⁻¹` in `mellin P.g s`
+  have step1 := mellin_comp_rpow P.g (-s) (-1)
+  simp_rw [abs_neg, abs_one, inv_one, one_smul, ofReal_neg, ofReal_one, div_neg, div_one, neg_neg,
+    rpow_neg_one, ← one_div] at step1
+  -- introduce a power of `t` to match the hypothesis `P.h_feq`
+  have step2 := mellin_cpow_smul (fun t ↦ P.g (1 / t)) (P.k - s) (-P.k)
+  rw [← sub_eq_add_neg, sub_right_comm, sub_self, zero_sub] at step2
+  simp_rw [step1] at step2
+  -- put in the constant `P.ε`
+  have step3 := mellin_const_smul (fun t ↦ (t : ℂ) ^ (-P.k : ℂ) • P.g (1 / t)) (P.k - s) P.ε
+  rw [step2] at step3
+  rw [← step3]
+  -- now the integrand matches `P.h_feq'` on `Ioi 0`, so we can apply `set_integral_congr`
+  refine set_integral_congr measurableSet_Ioi (fun t ht ↦ ?_)
+  simp_rw [P.h_feq' t ht, ← mul_smul]
+  -- some simple `cpow` arithmetic to finish
+  rw [cpow_neg, ofReal_cpow (le_of_lt ht)]
+  have : (t : ℂ) ^ (P.k : ℂ) ≠ 0 := by
+    simpa only [← ofReal_cpow (le_of_lt ht), ofReal_ne_zero] using (rpow_pos_of_pos ht _).ne'
+  field_simp [P.hε]
+
+end StrongFEPair
+
+namespace WeakFEPair
+
+variable (P : WeakFEPair E)
+
+/-!
+## Auxiliary results II: building a strong FE-pair from a weak FE-pair
+-/
+
+/-- Piecewise modified version of `f` with optimal asymptotics. We deliberately choose intervals
+which don't quite join up, so the function is `0` at `x = 1`, in order to maintain symmetry;
+there is no `good` choice of value at `1`. -/
+def f_modif : ℝ → E :=
+  (Ioi 1).indicator (fun x ↦ P.f x - P.f₀) +
+  (Ioo 0 1).indicator (fun x ↦ P.f x - (P.ε * ↑(x ^ (-P.k))) • P.g₀)
+
+/-- Piecewise modified version of `g` with optimal asymptotics. -/
+def g_modif : ℝ → E :=
+  (Ioi 1).indicator (fun x ↦ P.g x - P.g₀) +
+  (Ioo 0 1).indicator (fun x ↦ P.g x - (P.ε⁻¹ * ↑(x ^ (-P.k))) • P.f₀)
+
+lemma hf_modif_int :
+    LocallyIntegrableOn P.f_modif (Ioi 0) := by
+  have : LocallyIntegrableOn (fun x : ℝ ↦ (P.ε * ↑(x ^ (-P.k))) • P.g₀) (Ioi 0) := by
+    refine ContinuousOn.locallyIntegrableOn ?_ measurableSet_Ioi
+    refine ContinuousAt.continuousOn (fun x (hx : 0 < x) ↦ ?_)
+    refine (continuousAt_const.mul ?_).smul continuousAt_const
+    exact continuous_ofReal.continuousAt.comp (continuousAt_rpow_const _ _ (Or.inl hx.ne'))
```

Comment:
This looks like something `fun_prop` should be able to solve automatically at some point. Unfortunately, right now, it does not yet work...